R/AllClasses.R

```diff
@@ -0,0 +1,223 @@
+
+# Validation function ----
+
+.valid.Colormap <- function(object){
+  # To avoid later checking inside the app, this object should be stored in a
+  # slot of its parent SummarizedExperiment that would check its validity
+  # (e.g. all assay color maps should be present in the parent, etc.)
+  # Note that default color maps will be provided for elements
+  # of the parent that do not have a defined colormap.
+
+  errors <- c()
+
+  if (!all(sapply(object@assays, "typeof") == "character")){
+    errors <- c(errors, "Non-character vectors present in assay colormaps.")
+  }
+
+  if (!all(sapply(object@colData, "typeof") == "character")){
+    errors <- c(errors, "Non-character vectors present in assay colormaps.")
+  }
+
+  if (!all(sapply(object@rowData, "typeof") == "character")){
+    errors <- c(errors, "Non-character vectors present in assay colormaps.")
+  }
+
+  if (length(errors > 0)){
+    return(errors)
+  }
+
+  return(TRUE)
+}
+
+# ExperimentColorMap definition ----
+
+setClass("ExperimentColorMap",
+    contains="Vector",
+    representation(
+      assays="list", # continuous colormaps for assays
+      colData="list", # continuous or discrete colormaps for sample metadata
+      rowData="list" # continuous or discrete colormaps for gene metadata
+    ),
+    prototype(),
+  validity = .valid.Colormap
+)
+
+# Constructor ----
+
+#' \code{ExperimentColorMap} objects
+#'
+#' @name ExperimentColorMap
+#'
+#' @param assays List of color maps for \code{assays}.
+#' @param colData List of color maps for \code{colData}.
+#' @param rowData List of color maps for \code{rowData}.
+#' @param ... additional arguments passed on to the \code{ExperimentColorMap}
+#' constructor
+#'
+#' @section Accessors:
+#'
+#' In the following code snippets, \code{x} is an
+#' \code{ExperimentColorMap} object.
+#'
+#' \code{assayColorMap(x, "counts")}: Get an assay colormap.
+#'
+#' @export ExperimentColorMap
+#' @exportClass ExperimentColorMap
+#' @export assayColorMap
+#' @exportMethod assayColorMap
+#' @export colDataColorMap
+#' @export rowDataColorMap
+#' @aliases class:ExperimentColorMap ExperimentColorMap-class
+#' assayColorMap colDataColorMap rowDataColorMap
+#' assayColorMap,ExperimentColorMap,character-method
+#' assayColorMap,ExperimentColorMap,integer-method
+#' colDataColorMap,ExperimentColorMap,character-method
+#' colDataColorMap,ExperimentColorMap,integer-method
+#' rowDataColorMap,ExperimentColorMap,character-method
+#' rowDataColorMap,ExperimentColorMap,integer-method
+#'
+#' @examples
+#'
+#' # Constructor ----
+#'
+#' count_colors <- viridis::viridis(10)
```

Comment:
Is there an easy way to set the same colormap for all assays?

Reply:
We can think of a special call to the constructor for that.
Or maybe somehow allow the user to change the “default” colormap, currently hard coded

Reply:
I think you'd need another slot as there's no way for the color map object to know how many assays you have in total. To me, it seems like we need another two levels of slots:

- `all_assays`, etc.: contains two color maps for use for all assays (or all coldata, or all gene exprs). One color map is continuous, another is discrete.
- `global`: contains two color maps for use in everything, one continuous and one discrete.

More specific color maps are favoured, so `assays > all_assays > global >` our defaults.

Reply:
I'm glad to see you embrace the `ExperimentColorMap` thinking :)
That's all made fairly easy by the formal class structure.